source/grid/grid_tools.cc

```diff
@@ -1458,6 +1458,145 @@ namespace GridTools
+                                 const Point<spacedim>                                                          &p,
+                                 const typename MeshType<dim, spacedim>::active_cell_iterator                   &current_cell,
+                                 const std::vector<std::set<typename MeshType<dim,spacedim>::active_cell_iterator > > &vertex_to_cells,
+                                 const std::vector<std::vector<Tensor<1,spacedim> > >                            &vertex_to_cell_centers,
+                                 const std::vector<bool>                                                         &marked_vertices)
+  {
+    std::pair<typename MeshType<dim, spacedim>::active_cell_iterator, Point<dim> > cell_and_position;
+
+    // Create a corresponding map of vectors from vertex to cell center
+    std::vector<unsigned int> neighbor_permutation;
+
+    bool found_cell = false;
+
+    const unsigned int closest_vertex = get_closest_vertex_of_cell(current_cell, p);
+    Tensor<1,spacedim> vertex_to_point = p - current_cell->vertex(closest_vertex);
+    vertex_to_point /= vertex_to_point.norm();
+
+    const unsigned int closest_vertex_index = current_cell->vertex_index(closest_vertex);
+    const unsigned int n_neighbor_cells = vertex_to_cells[closest_vertex_index].size();
+
+    neighbor_permutation.resize(n_neighbor_cells);
+    for (unsigned int i=0; i<n_neighbor_cells; ++i)
+      neighbor_permutation[i] = i;
+
+    std::sort(neighbor_permutation.begin(),
+              neighbor_permutation.end(),
+              std::bind(&compair_point_association<dim>,
+                        _1,
+                        _2,
+                        std::cref(vertex_to_point),
+                        std::cref(vertex_to_cell_centers[closest_vertex_index])));
+
+    // Search all of the cells adjacent to the closest vertex of the previous cell
+    // Most likely we will find the particle in them.
+    for (unsigned int i=0; i<n_neighbor_cells; ++i)
+      {
+        try
+          {
+            typename std::set<typename Triangulation<dim>::active_cell_iterator>::const_iterator cell = vertex_to_cells[closest_vertex_index].begin();
+            std::advance(cell,neighbor_permutation[i]);
+            const Point<dim> p_unit = mapping.transform_real_to_unit_cell(*cell, p);
+            if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+              {
+                cell_and_position.first = *cell;
+                cell_and_position.second = p_unit;
+                found_cell = true;
+                break;
+              }
+          }
+        catch (typename Mapping<dim>::ExcTransformationFailed &)
+          {}
+      }
+
+
+    if (!found_cell)
+      {
+        // The particle is not in a neighbor of the old cell.
+        // Look for the new cell in the whole local domain.
+        // This case is rare.
```

Comment:
I would extend this to say: "This case is rare if the vertex is in cell_guess or in its neighbors."